### PR TITLE
SCM function handling by GOSUB return.

### PR DESCRIPTION
**Pull request by MiranDMC**

### CODE REVIEW

File: source/CCustomOpcodeSystem.cpp [hunk 1]
```diff
@@ -850,6 +854,22 @@ namespace CLEO
 	/*						Opcode definitions								*/
 	/************************************************************************/
 
+	OpcodeResult __stdcall CCustomOpcodeSystem::opcode_0051(CRunningScript* thread) // GOSUB return
+	{
+		if (thread->SP == 0 && thread->IsCustom() && !IsLegacyScript(thread)) // CLEO5 - allow use of GOSUB `return` to exit cleo calls too
+		{
+			return opcode_0AB2(thread); // try CLEO's function return
```
Comment (x87) [inline source/CCustomOpcodeSystem.cpp:861]: 0051 isn't var args opcode. please use CleoReturnGeneric directly

Comment (x87) [reply source/CCustomOpcodeSystem.cpp:861]:
```
return GetInstance().OpcodeSystem.CleoReturnGeneric(0x0051, thread, false);
```

Comment (MiranDMC) [reply source/CCustomOpcodeSystem.cpp]: updated

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: is thread->IsCustom() check necessary? we might be able to use it in main.scm too

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: how about

if sp ==0 {
 get scm func
 if scm == nullptr {
   // error, no gosub nor function
  // return is used outside of subroutine (gosub or scm call), suspended
 }
 call returnGeneric
}
call 0051

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: hmm, if single `return` can exit from gosub subroutine, then `return true/false` should be able too. but then we get back to square one, as I wanted to keep an ability to exit from any nested gosub just by returning from the function.

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: or probably not. `return true/false` in gosub should be an error

Comment (MiranDMC) [reply source/CCustomOpcodeSystem.cpp]: > how about showing a custom message when `return` is found outside of any gosub or scm function?
CleoReturnGeneric has it's own error message.

Comment (MiranDMC) [reply source/CCustomOpcodeSystem.cpp]: > is thread->IsCustom() check necessary? we might be able to use it in main.scm too

https://github.com/cleolibrary/CLEO5/blob/e6ab655c46ac4481484866ef6b4dbf4b9615b276/source/CScriptEngine.cpp#L249

It itself contains IsCustom check.

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: if we should consider a gosub call as a scopeless, parameter-less variation of scm function. then using both `return` and `return true/false` can be justifiable in gosub.
```

what do you think of this:

|              | 0051: return | 0AB2: 0 | 0AB2: values | 2002: flag | 2002: flag values | 2003 |
| ------------ | ------------ | ------- | ------------ | ---------- | ----------------- | ---- |
| gosub        | OK           | ERROR\* | ERROR\*      | OK         | ERROR             | OK   |
| call 0       | OK           | OK      | ERROR        | OK         | ERROR             | OK   |
| call outputs | ERROR        | ERROR   | OK           | ERROR      | OK                | OK   |

- backward compatibility with CLEO4

Comment (MiranDMC) [reply source/CCustomOpcodeSystem.cpp]: so `return` would be alias to '0AB2: cleo_return 0` if called inside cleo function?
`return true/false` -> `cleo_return_with true/false`

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: return true/false does not return any arguments, it's condition flag. 

that's the point. even if you don't care about condition result you are now forced to provide one in order to use `return` in a function now. I used something like `return true` everywhere in cleo testing library.

now, with this change you should be able to use just `return` and get consistent results with `return true/false` .

Comment (x87) [reply source/CCustomOpcodeSystem.cpp]: so a function with no output and gosub will be treated in the same way. you can exit both using `return` or `return true/false`, depending on whether or not you care about condition flag.

`return true/false values` syntax can only be used in functions returning something.

Comment (MiranDMC) [reply source/CCustomOpcodeSystem.cpp]: Updated.
What to do with main scm is still debatable. In theory legacy mode can be enabled in cleo.ini, but then the setting will be shared between main game and all mod packs. scm3 scm4 would make sense.

**Comment by x87** [comment]
appeared to be working. tests and sanny updated to support single return in functions.

**Review by x87 — APPROVED** [review]

**Pull request merged**